### PR TITLE
Add MooseServer definitions, formatting, and VS Code updates

### PR DESCRIPTION
**Pull request by brandonlangley**
This includes the following updates with the MOOSE language server.

 - full document formatting capability

 - goto-definition capability including
   - look up of definitions from the current or !included input files
   - jump to source file where the object was registered for types
   - send empty response if no file exists at path for source code

 - fix-ups from feedback via VS Code
   - initialize with the current set of supported server capabilities
   - strip arbitrary URI schemes fixing diagnostic line and column
   - remove completion description lead spaces and (REQUIRED)
   - set SymbolKind value on document symbols for outline icons
   - set CompletionItemKind values used by completion list icons

Here is NEAMS Workbench demonstrating definitions and formatting.

https://github.com/idaholab/moose/assets/21372118/f127cf3b-f7a1-4927-9c38-5e5ab8943a0a

refs #22766

### CODE REVIEW

**Comment by loganharbour** [comment]
Remove the draft status when you're ready for a review and I'll take a look.

**Comment by brandonlangley** [comment]
@loganharbour -

This adds both go-to-definitions and input-formatting capabilities to the MooseServer.

I had to make a change to WASP for document formatting which is still awaiting merge.

After that's merged, I'll update the WASP submodule here and remove the draft status.

**Comment by brandonlangley** [comment]
@loganharbour -

This is ready for review now and is a much smaller change-set than the previous PRs split into these four commits:

1. Goto-Definitions capability implementation and unit testing

2. Input-Formatting capability implementation and unit testing

3. Needed update to WASP submodule and Conda YAML files

4. Bump in versioner_hashes.yaml as a result of WASP update

Here's a short video showing the Goto-Definitions then Input-Formatting features in Workbench using this branch.

https://github.com/idaholab/moose/assets/21372118/960724e6-436c-4e8f-bbeb-cbc1321318e4

**Comment by dschwen** [comment]
We indent .i files with two spaces. Is that configurable? It should be the default to avoid creating large diffs when editing files that have been formatted with `hit format`.

**Comment by brandonlangley** [comment]
@dschwen -

> We indent .i files with two spaces. Is that configurable? It should be the default to avoid creating large diffs when editing files that have been formatted with `hit format`.

Right, the `hit::Formatter::indent_string` is defaulted to `"  "` which I think is a good formatting indention default.

The changes here don't modify that default or make any changes at all to the `hit::Formatter`.

A `hit::Formatter` is just used by the `MooseServer` when a client requests formatting over the language server protocol.

And the language server protocol document formatting request actually [requires](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#formattingOptions) the `tabSize` parameter to be specified.

So any time a client / IDE requests document formatting, they will also always have to supply this `tabSize`  parameter.

The `MooseServer` logic that handles the formatting request simply uses the integer value of that required parameter:

```cpp
  hit::Formatter formatter("cfg", "[format]indent_string='" + std::string(tab_size, ' ') + "'[]");
  std::string document_formatted = formatter.format(document_path, document_text);
```

It seems like an odd choice in the protocol that clients are forced to pick a formatting indention size as mentioned here:

https://github.com/microsoft/language-server-protocol/issues/1166#issuecomment-957325352

It would be nice if it allowed a client to defer that decision to the language server that is actually doing the formatting.

In the `MooseServerTest`, I set `tab_size = 4` since the default indention size in the `hit::Formatter` is two spaces.

And I wanted to test formatting with a value different from two to ensure the requested size is [actually being applied](https://github.com/idaholab/moose/pull/25305/commits/da6109a6fa8eeed9469370e6a0c5d0bd3a7177ca#diff-e861ff6bdefa1faecc5dfe43d534b0008bfeabd7d6895ac6dad76c4cd088c0faR1570-R1628).

**Comment by dschwen** [comment]
Can you show that this works with actions? E.g. variables generated by actions. It would be nice to retain the current plugin's behavior of jumping to source code for `type` definitions.

**Comment by brandonlangley** [comment]
@dschwen 

> Can you show that this works with actions? E.g. variables generated by actions. It would be nice to retain the current plugin's behavior of jumping to source code for `type` definitions.

Right now the Go-to-definition implementation provides definitions within the input file for referential lookup values, e.g.,

```
01    [Variables]
02      [u]
03        family = MONOMIAL
04        order = CONSTANT
05      []
06    []
07    [BCs]
08      [all]
09        type = VacuumBC
10        boundary = 'left right'
11        variable = u
12      []
13    []
```

with the cursor on the value `u` on line 11, `Go to definition` moves the cursor to select the definition of `[u]` on line 2.

So this capability does not currently provide any response locations when used on the value of a  block `type` parameter.

I like this idea though and I see how the `getLineInfo()` calls are being used to add `file_info` to the JSON syntax tree.

Since `type` parameter values are unused in this feature now, I can add this without conflicting with any current behavior.

The one question I have is with unit testing a MooseServer response since locations consist of file-path and line-number.

Is it okay to test a specific location since the line-number will be invalid if even one line gets added that source code file?

**Comment by brandonlangley** [comment]
@dschwen -

This is finished now and ready for your review plus any further testing in VS Code that you would like - it includes:

 - full document formatting capability

 - goto-definition capability including
   - look up of definitions from the current or !included input files
   - jump to source file where the object was registered for types
   - send empty response if no file exists at path for source code

 - fix-ups via your VS Code feedback
   - initialize with the current set of supported server capabilities
   - strip arbitrary URI schemes fixing diagnostic line and column
   - remove completion description lead spaces and (REQUIRED)
   - set SymbolKind value on document symbols for outline icons
   - set CompletionItemKind values used by completion list icons

I think more complex pieces from https://github.com/idaholab/moose/issues/25787 like action generated names or block restrictions can wait on a later PR.

Regarding SymbolKind outline icons and CompletionItemKind list icons, I found this image of the full VS Code set:

CompletionItemKind-SymbolKind-VSCode-Icons

**Comment by brandonlangley** [comment]
@dschwen -

FYI, I'm planning to wait until after your review and testing to rebase this branch and fix the merge conflicts if that's okay.

**Comment by brandonlangley** [comment]
@loganharbour -

I just rebased this fixing the merge conflicts and dropped the `wasp`, `conda`, and `versioner_hashes` commits.

This is finished and ready to merge at your convenience.

File: framework/src/base/MooseServer.C
Comment (loganharbour) [inline framework/src/base/MooseServer.C:770]: Does this link object in input -> source file?

**Review by loganharbour — APPROVED** [review]
Looks good to me. I think that a @dschwen once-through is a good idea, though.

File: framework/src/base/MooseServer.C
Comment (brandonlangley) [reply framework/src/base/MooseServer.C:770]: @loganharbour -

Yes, this makes `type=ObjectName` work like goto-definition in @dschwen's VSCode plugin as he [requested](https://github.com/idaholab/moose/pull/25305#issuecomment-1749286453).

It returns the path to the `ObjectName.C` source code file where the object was registered for the GUI to use.

I talked to him previously about what to do when the source file does not exist as in executable-only installs.

And it was decided the best thing to do in that situation is send an empty response, so this check does that:

```cpp
    // return without any definition locations added if file cannot be read
    if (!file_line_info.isValid() ||
        !MooseUtils::checkFileReadable(file_line_info.file(), false, false, false))
```

This type --> source scenario is unit tested in the `DefinitionObjectTypeSource` case of `MooseServerTest`.

**Comment by brandonlangley** [comment]
@dschwen -

I just rebased this and fixed the `MooseServerTest` merge conflict so this is ready for you to look at again and merge.

**Review by dschwen — APPROVED** [review]

**Pull request merged**
